### PR TITLE
Assert expected number of sources in catalog config (#5081)

### LINKED ISSUE #5081: Assert expected number of sources in catalog config
We didn't catch an operator error, forgetting to include `pop` for a snapshot in the dcp25 catalog:

https://humancellatlas.slack.com/archives/C9XD6L0AD/p1679931329270429?thread_ts=1678825156.333889&cid=C9XD6L0AD

https://ucsc-gi.slack.com/archives/C705Y6G9Z/p1679933593255449

---

- [ ] Security design review completed; the Resolution of this issue does **not** …
  - [ ] … affect authentication; for example:
    - OAuth 2.0 with the application (API or Swagger UI)
    - Authentication of developers with Google Cloud APIs
    - Authentication of developers with AWS APIs
    - Authentication with a GitLab instance in the system
    - Password and 2FA authentication with GitHub
    - API access token authentication with GitHub
    - Authentication with 
  - [ ] … affect the permissions of internal users like access to
    - Cloud resources on AWS and GCP
    - GitLab repositories, projects and groups, administration
    - an EC2 instance via SSH
    - GitHub issues, pull requests, commits, commit statuses, wikis, repositories, organizations
  - [ ] … affect the permissions of external users like access to
    - TDR snapshots
  - [ ] … affect permissions of service or bot accounts
    - Cloud resources on AWS and GCP
  - [ ] … affect audit logging in the system, like
    - adding, removing or changing a log message that represents an auditable event
    - changing the routing of log messages through the system
  - [ ] … affect monitoring of the system
  - [ ] … introduce a new software dependency like
    - Python packages on PYPI
    - Command-line utilities
    - Docker images
    - Terraform providers
  - [ ] … add an interface that exposes sensitive or confidential data at the security boundary
  - [ ] … affect the encryption of data at rest
  - [ ] … require persistence of sensitive or confidential data that might require encryption at rest
  - [ ] … require unencrypted transmission of data within the security boundary
  - [ ] … affect the network security layer; for example by 
    - modifying, adding or removing firewall rules
    - modifying, adding or removing security groups
    - changing or adding a port a service, proxy or load balancer listens on
- [ ] Documentation on any unchecked boxes is provided in comments below

### PR DESCRIPTION
**Pull request by achave11-ucsc**
Connected issues: #5081

### Author

- [x] PR is a draft
- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] PR title references all connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue <sub>or comment in PR explains why they're different</sub>
- [x] For each connected issue, there is at least one commit whose title references that issue
- [x] PR is connected to all connected issues via ZenHub
- [x] PR description links to connected issues
- [x] Added `partial` label to PR <sub>or this PR completely resolves all connected issues</sub>

<sup>1</sup> when the issue title describes a problem, the corresponding PR
title is `Fix: ` followed by the issue title

### Author (reindex, API changes)

- [x] Added `r` tag to commit title <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR <sub>or this PR does not require reindexing</sub>
- [x] Added `a` (compatible changes) or `A` (incompatible ones) tag to commit title <sub>or this PR does not modify the Azul service API</sub>
- [x] Added `API` label to connected issues <sub>or this PR does not modify the Azul service API</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain <sub>or this PR is not chained to another PR</sub>
- [x] Added `base` label to the blocking PR <sub>or this PR is not chained to another PR</sub>
- [x] Added `chained` label to this PR <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR <sub>or this PR does not require upgrading</sub>

### Author (operator tasks)

- [x] Added checklist items for additional operator tasks <sub>or this PR does not require additional tasks</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues <sub>or the `prod` branch has no temporary hotfixes for any connected issues</sub>

### Author (before every review)

- [x] Rebased PR branch on `develop`, squashed old fixups
- [x] Ran `make requirements_update` <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR <sub>or this PR does not touch requirements*.txt</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>

### Peer reviewer (after requesting changes)

Uncheck the *Author (before every review)* checklists.

### Peer reviewer (after approval)

- [x] PR is not a draft
- [x] Ticket is in *Review requested* column
- [x] Requested review from primary reviewer
- [x] Assigned PR to primary reviewer

### Primary reviewer (after requesting changes)

Uncheck the *before every review* checklists. Update the `N reviews` label.

### Primary reviewer (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved ticket to *Approved* column
- [x] Assigned PR to current operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear <sub>or all connected issues are labeled `no demo`</sub>
- [x] Squashed PR branch and rebased onto `develop`
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Pushed PR branch to GitLab `dev` and added `sandbox` label <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvilbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Started reindex in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved connected issues to Merged column in ZenHub
- [x] Pushed merge commit to GitHub

### Operator (after pushing the merge commit)

- [x] Shortened the PR chain <sub>or this PR is not labeled `base`</sub>
- [x] Pushed merge commit to GitLab `dev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed merge commit to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes on GitLab `dev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `dev`<sup>1</sup>
- [x] Build passes on GitLab `anvildev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `anvildev`<sup>1</sup>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`

<sup>1</sup> When pushing the merge commit is skipped due to the PR being
labelled `no sandbox`, the next build triggered by a PR whose merge commit *is*
pushed determines this checklist item.

### Operator (reindex)

- [x] Deleted unreferenced indices in `dev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvildev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Started reindex in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `dev` deployment <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `anvildev` deployment <sub>or this PR does not require reindexing</sub>

### Operator

- [x] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

File: deployments/anvilbox/environment.py
Comment (nadove-ucsc) [inline deployments/anvilbox/environment.py]: Why was this change made to the dictionary keys? And why is the splitting different here than in the `anvildev` environment file, given that they use the same set of snapshots?

Comment (nadove-ucsc) [reply deployments/anvilbox/environment.py]: Assuming the logic is correct, I think this could be better expressed as 

```suggestion
    project  = '_'.join(snapshot.split('_')[1:-3])
```

**Review by nadove-ucsc — CHANGES_REQUESTED** [review]

File: deployments/anvilbox/environment.py
Comment (achave11-ucsc) [reply deployments/anvilbox/environment.py]: >Why was this change made to the dictionary keys?

For it to be similar in convention to the HCA dictionary keys. Reference https://github.com/DataBiosphere/azul/issues/5081#issuecomment-1485531487.

>And why is the splitting different here than in the `anvildev` environment file, given that they use the same set of snapshots?

It was an oversight.

Your suggestion wasn't implemented verbatim, but a flavor of it was.

File: deployments/anvilprod/environment.py
Comment (nadove-ucsc) [inline deployments/anvilprod/environment.py]: Should match `anvildev` and `anvilbox`

File: deployments/anvilbox/environment.py
Comment (nadove-ucsc) [reply deployments/anvilbox/environment.py]: > For it to be similar in convention to the HCA dictionary keys. Reference https://github.com/DataBiosphere/azul/issues/5081#issuecomment-1485531487.

I can see how this is a good idea, but I don't understand how that comment motivates this change.

File: deployments/anvilbox/environment.py
Comment (nadove-ucsc) [inline deployments/anvilbox/environment.py]: This doesn't really reflect the intent of my suggestion. My main issue with this section is how the `project` variable is initialized with parts of the string that aren't part of the project name. If you feel it's important to use unpacking here, then I suggest an alternative compromise between your initial approach and the one I suggested:

```suggestion
    env, *project, _, _, _ = snapshot.split('_')
    project = '_'.join(project)
```

Comment (nadove-ucsc) [reply deployments/anvilbox/environment.py]: Also applies to `anvildev` and `anvilprod`

**Review by nadove-ucsc — CHANGES_REQUESTED** [review]

File: deployments/anvilbox/environment.py
Comment (achave11-ucsc) [reply deployments/anvilbox/environment.py]: I implemented your previous suggestion, the intend is clearer like you said.

File: deployments/anvilbox/environment.py
Comment (achave11-ucsc) [reply deployments/anvilbox/environment.py]: I brought it up during PL and @hannes-ucsc and I sorted it out and came up with that decision.

**Review by nadove-ucsc — DISMISSED** [review]

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
```diff
Index: deployments/sandbox/environment.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/deployments/sandbox/environment.py b/deployments/sandbox/environment.py
--- a/deployments/sandbox/environment.py	(revision 0a141e8497cb7f5f8f75e9362139e3df5af91d72)
+++ b/deployments/sandbox/environment.py	(date 1680890389688)
@@ -52,6 +52,7 @@
     assert list(result.keys()) == sorted(result.keys()), 'input not sorted'
     return result
 
+
 def mklist(catalog: dict[str, str]) -> list[str]:
     return list(filter(None, catalog.values()))
 
Index: deployments/anvilbox/environment.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/deployments/anvilbox/environment.py b/deployments/anvilbox/environment.py
--- a/deployments/anvilbox/environment.py	(revision 0a141e8497cb7f5f8f75e9362139e3df5af91d72)
+++ b/deployments/anvilbox/environment.py	(date 1680890463376)
@@ -52,6 +52,7 @@
     assert list(result.keys()) == sorted(result.keys()), 'input not sorted'
     return result
 
+
 def mklist(catalog: dict[str, str]) -> list[str]:
     return list(filter(None, catalog.values()))
 
Index: deployments/dev/environment.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/deployments/dev/environment.py b/deployments/dev/environment.py
--- a/deployments/dev/environment.py	(revision 0a141e8497cb7f5f8f75e9362139e3df5af91d72)
+++ b/deployments/dev/environment.py	(date 1680890600422)
@@ -40,6 +40,7 @@
     assert list(result.keys()) == sorted(result.keys()), 'input not sorted'
     return result
 
+
 def mklist(catalog: dict[str, str]) -> list[str]:
     return list(filter(None, catalog.values()))
 
Index: deployments/prod/environment.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/deployments/prod/environment.py b/deployments/prod/environment.py
--- a/deployments/prod/environment.py	(revision 0a141e8497cb7f5f8f75e9362139e3df5af91d72)
+++ b/deployments/prod/environment.py	(date 1680890389685)
@@ -286,7 +286,7 @@
     mksrc('datarepo-590e9f21', 'hca_prod_faeedcb0e0464be7b1ad80a3eeabb066__20220119_dcp2_20220121_dcp12', 62),
 ]))
 
-dcp13_sources = mkdict(dcp12_sources, 208 , mkdelta([
+dcp13_sources = mkdict(dcp12_sources, 208, mkdelta([
     mksrc('datarepo-c8f9ec5d', 'hca_prod_03c6fce7789e4e78a27a664d562bb738__20220110_dcp2_20220214_dcp13', 1531),
     mksrc('datarepo-991fac12', 'hca_prod_05657a599f9d4bb9b77b24be13aa5cea__20220110_dcp2_20220214_dcp13', 186),
     mksrc('datarepo-b185fd85', 'hca_prod_05be4f374506429bb112506444507d62__20220107_dcp2_20220214_dcp13', 1545),
@@ -392,7 +392,7 @@
     mksrc('datarepo-b60aabf3', 'hca_prod_fde199d2a8414ed1aa65b9e0af8969b1__20220330_dcp2_20220330_dcp15', 185),
 ]))
 
-dcp16_sources = mkdict(dcp15_sources, 250 , mkdelta([
+dcp16_sources = mkdict(dcp15_sources, 250, mkdelta([
     mksrc('datarepo-c531f177', 'hca_prod_0562d2ae0b8a459ebbc06357108e5da9__20220330_dcp2_20220519_dcp16', 17),
     mksrc('datarepo-589be0ea', 'hca_prod_0b29914025b54861a69f7651ff3f46cf__20220519_dcp2_20220519_dcp16', 82),
     mksrc('datarepo-a584f228', 'hca_prod_16cd67912adb4d0f82220184dada6456__20220519_dcp2_20220519_dcp16', 395),
Index: deployments/anvildev/environment.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/deployments/anvildev/environment.py b/deployments/anvildev/environment.py
--- a/deployments/anvildev/environment.py	(revision 0a141e8497cb7f5f8f75e9362139e3df5af91d72)
+++ b/deployments/anvildev/environment.py	(date 1680890389680)
@@ -40,6 +40,7 @@
     assert list(result.keys()) == sorted(result.keys()), 'input not sorted'
     return result
 
+
 def mklist(catalog: dict[str, str]) -> list[str]:
     return list(filter(None, catalog.values()))
 
```

And this as a separate commit

```diff
Index: Makefile
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Makefile b/Makefile
--- a/Makefile	(revision 0a141e8497cb7f5f8f75e9362139e3df5af91d72)
+++ b/Makefile	(date 1680890926929)
@@ -194,6 +194,8 @@
                                 $(project_root)/test \
                                 $(project_root)/lambdas/{layer,indexer,service}/app.py \
                                 $(project_root)/.flake8/azul_flake8.py \
+                                $(project_root)/environment.py \
+                                $(project_root)/deployments/*/environment.py \
                                 $$(find $(project_root)/terraform{,/gitlab,/shared} \
                                         $(project_root)/lambdas/{indexer,service}{,/.chalice} \
                                         $(project_root)/.github \
Index: environment.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/environment.py b/environment.py
--- a/environment.py	(revision 0a141e8497cb7f5f8f75e9362139e3df5af91d72)
+++ b/environment.py	(date 1680890961694)
@@ -424,7 +424,6 @@
         #
         'azul_chalice_bin': '{project_root}/bin/wheels/runtime',
 
-
         # Stop `pip` from nagging us about updates. We update pip regularly like
         # any other dependency. There is nothing special about `pip` that would
         # warrant the distraction.
@@ -508,7 +507,7 @@
         # develop branch, 0 otherwise. Personal deployments have this set to 0.
         #
         'AZUL_IS_SANDBOX': '0',
-        
+
         # A list of names of AWS IAM roles that should be given permission to
         # manage incidents with AWS support as defined in CIS rule 1.20:
         #
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Looks good but this will break when combined with the changes from PR #5149 for #5134. Please wait until that lands and re-request review.

**Review by hannes-ucsc — DISMISSED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Pull request merged**
